### ISSUE TITLE
User is allowed to see columns of PostgreSQL table without SELECT privilege

### ISSUE THREAD

**Issue by kokosing** [trinodb/trino#6947]
Below test code reproduces the issue.

```
            String accessDeniedErrorMessageRegexp = "(?s)ERROR: permission denied for schema tpch.*";
            assertQueryFails(userSession, "SELECT * FROM nation", accessDeniedErrorMessageRegexp);
            assertQuery(
                    userSession,
                    "DESC nation",
                    "VALUES " +
                            "('nationkey', 'bigint', '', '')," +
                            "('name', 'varchar(25)', '', '')," +
                            "('regionkey', 'bigint', '', '')," +
                            "('comment', 'varchar(152)', '', '')");
            assertQuery(
                    userSession,
                    "SHOW CREATE TABLE nation",
                    "VALUES " +
                            "'CREATE TABLE postgresql.tpch.nation (\n" +
                            "   nationkey bigint,\n" +
                            "   name varchar(25),\n" +
                            "   regionkey bigint,\n" +
                            "   comment varchar(152)\n" +
                            ")'");
                            ```

**Comment by kokosing** [comment]
See https://github.com/trinodb/trino/pull/6948

**Comment by findepi** [comment]
PostgreSQL connector has no access control, it this specific to PostgreSQL?
or is it about SystemAccessControl?

**Comment by kokosing** [comment]
It is specific to PostgreSQL. No access control in Trino were used.

**Comment by kokosing** [comment]
Once @electrum said:

This a complicated subject. As to how information_schema should work, you can see the definitions in part 11 of the SQL standard (ISO 9075-11). The columns view includes this

```
WHERE (C.TABLE_CATALOG, C.TABLE_SCHEMA, C.TABLE_NAME, C.COLUMN_NAME) IN
   (SELECT CP.TABLE_CATALOG, CP.TABLE_SCHEMA, CP.TABLE_NAME, CP.COLUMN_NAME
    FROM DEFINITION_SCHEMA.COLUMN_PRIVILEGES AS CP
    WHERE (CP.GRANTEE IN ('PUBLIC', CURRENT_USER) 
       OR CP.GRANTEE IN (SELECT ROLE_NAME FROM ENABLED_ROLES)))
```
   
The tables view has a similar filter using table_privileges. This means that users only see objects for which they have privileges.

Based on the above, this seems to mean that PostgreSQL are wrong according to the SQL specification. 
[...]

**Comment by findepi** [comment]
> Based on the above, this seems to mean that PostgreSQL are wrong according to the SQL specification.

PostgreSQL or PostgreSQL connector?

**Comment by kokosing** [comment]
PostgreSQL, you can easily reproduce the issue connecting directly to PostgreSQL when you select from information_schema there. It is not only related to columns, but to tables too. It is the way PostgreSQL behave today and same behavior is exposed via Trino. So I am not 100% sure if it is actually a bug.

**Comment by kokosing** [comment]
This is a bit related to https://github.com/trinodb/trino/pull/6916, so if we like we could possibly improve PostgreSQL to throw access denied when accessing information about table columns.

**Comment by electrum** [comment]
I don’t understand this at all. Why would we get access denied from Trino if there is no access control?

I think you took my comment out of context. I was describing what the SQL standard says. We don’t enforce SQL standard access rules in Trino — it’s up to the access control to decide to do that.

If you are using impersonation to PostgreSQL, then you would get whatever authorization behavior they implement. We shouldn’t try to implement something more restrictive in the connector.

**Comment by kokosing** [comment]
> I think you took my comment out of context. I was describing what the SQL standard says. 

That is my point. I took it to only to mention what SQL standard says. I didn't say that we should now enforce it to connectors that are not SQL standard compliant here.

> We shouldn’t try to implement something more restrictive in the connector.

Thanks for your opinion. That matches my understanding.

Motivation behind this issue is that we noticed this behavior so I wanted to discuss it here to understand if it is actually a bug or not. See https://github.com/trinodb/trino/issues/6947#issuecomment-781944284.

I am closing the issue as everything is as it is meant to be.
